Mathlib/RingTheory/RingHom/Locally.lean

```diff
@@ -0,0 +1,175 @@
+  rw [this, RingHom.comp_assoc]
+  exact hP.left _ _ (hf i)
+
+/-- Equivalence variant of `locally_of_exists`. This is sometimes easier to use, if the
+`IsLocalization.Away` instance can't be automatically inferred. -/
+lemma locally_iff_exists (hP : RespectsIso P) (f : R →+* S) :
+    Locally P f ↔ ∃ (ι : Type u) (s : ι → S) (_ : Ideal.span (Set.range s) = ⊤) (Sₜ : ι → Type u)
+      (_ : (i : ι) → CommRing (Sₜ i)) (_ : (i : ι) → Algebra S (Sₜ i))
+      (_ : (i : ι) → IsLocalization.Away (s i : S) (Sₜ i)),
+      ∀ i, P ((algebraMap S (Sₜ i)).comp f) :=
+  ⟨fun ⟨s, hsone, hs⟩ ↦ ⟨s, fun t : s ↦ (t : S), by simpa, fun t ↦ Localization.Away (t : S),
+      inferInstance, inferInstance, inferInstance, fun t ↦ hs t.val t.property⟩,
+    fun ⟨ι, s, hsone, Sₜ, _, _, hislocal, hs⟩ ↦ locally_of_exists hP f s hsone Sₜ hs⟩
+
+/-- In the definition of `Locally` we may replace `Localization.Away` with an arbitrary
+algebra satisfying `IsLocalization.Away`. -/
+lemma locally_iff_isLocalization (hP : RespectsIso P) (f : R →+* S) :
+    Locally P f ↔ ∃ (s : Finset S) (_ : Ideal.span (s : Set S) = ⊤),
+      ∀ t ∈ s, ∀ (Sₜ : Type u) [CommRing Sₜ] [Algebra S Sₜ] [IsLocalization.Away t Sₜ],
+      P ((algebraMap S Sₜ).comp f) := by
+  rw [locally_iff_finite P f]
+  refine ⟨fun ⟨s, hsone, hs⟩ ↦ ⟨s, hsone, fun t ht Sₜ _ _ _ ↦ ?_⟩, fun ⟨s, hsone, hs⟩ ↦ ?_⟩
+  · let e : Localization.Away t ≃+* Sₜ :=
+      (IsLocalization.algEquiv (Submonoid.powers t) _ _).toRingEquiv
+    have : algebraMap S Sₜ = e.toRingHom.comp (algebraMap S (Localization.Away t)) :=
+      RingHom.ext (fun x ↦ (AlgEquiv.commutes (IsLocalization.algEquiv _ _ _) _).symm)
+    rw [this, RingHom.comp_assoc]
+    exact hP.left _ _ (hs t ht)
+  · exact ⟨s, hsone, fun t ht ↦ hs t ht _⟩
+
+/-- If `f` satisfies `P`, then in particular it satisfies `Locally P`. -/
+lemma locally_of (hP : RespectsIso P) (f : R →+* S) (hf : P f) : Locally P f := by
+  use {1}
+  let e : S ≃+* Localization.Away (1 : S) :=
+    (IsLocalization.atUnits S (Submonoid.powers 1) (by simp)).toRingEquiv
+  simp only [Set.mem_singleton_iff, forall_eq, Ideal.span_singleton_one, exists_const]
+  exact hP.left f e hf
+
+/-- If `P` is local on the target, then `Locally P` coincides with `P`. -/
+lemma locally_eq_of_localizationSpanTarget (hPi : RespectsIso P)
+    (hPs : OfLocalizationSpanTarget P) {R S : Type u} [CommRing R] [CommRing S] (f : R →+* S) :
+    Locally P f ↔ P f :=
+  ⟨fun ⟨s, hsone, hs⟩ ↦ hPs f s hsone (fun a ↦ hs a.val a.property), locally_of hPi f⟩
+
+section OfLocalizationSpanTarget
+
+private noncomputable abbrev locallyAuxFun (s : Set S) (t : (r : s) → Set (Localization.Away r.val))
+    (p : (a : s) × t a) : S :=
+  p.1 * (IsLocalization.Away.sec p.1.1 p.2.1).1
+
+private lemma span_range_locallyAuxFun_eq_top {s : Set S} (hsone : Ideal.span s = ⊤)
+    {t : (r : s) → Set (Localization.Away r.val)} (htone : ∀ (r : s), Ideal.span (t r) = ⊤) :
+    Ideal.span (Set.range (locallyAuxFun s t)) = ⊤ := by
```

Comment:
I think they could be useful elsewhere and they should not be private (and potentially live in a different file).

Reply:
I gave them proper names and un`private`d. Not sure where these should go though, any suggestions? For now I just moved them to the top of this file in their own section.

Reply:
What does find-home say? Maybe `Mathlib/RingTheory/Localization/Away/Basic.lean` near `Away.sec`?

Reply:
`find_home` says `RingHom/Locally`. `Away.Basic` does not work, because it needs `Mathlib.RingTheory.Localization.Submodule`.

Reply:
You could open a new file under `Mathlib/RingTheory/Localization/Away/`, but I'm also okay with it being in this file. Please do what you see fit.

Reply:
Okay, I made a new file `RingTheory.Localization.Away.Lemmas`.